### PR TITLE
Make ArtifactTable.js sortable by clicking each column; force nowrap on buttons for small width screens

### PR DESCRIPTION
**Pull request by jsthomas1**
Users may sort in ascending or descending order by clicking the column name within the table. Sort is available four parameters of an Artifact object: Lexicographical order (name), Version String, Date Modified, and Date Created.

Regex is run against artifact version string before submission is allowed.

### CODE REVIEW

**Review by cmoesel — CHANGES_REQUESTED** [review]
Thank you for this PR, @jsthomas1!  This is great (and quick) work!  I've taken it for a spin and confirm that the basic functionality works well.  Nicely done!

I just have a few requests / comments (which are probably mostly a result of me not giving you the full requirements -- sorry!):

1. While we do recommend the APR versioning scheme, we intentionally do not enforce it.  This is because the CQL specification does not require it -- and some organizations have their own versioning scheme (for example, encoded dates like `20210312`). So... I think we'll need to back out that extra validation.

2. Related to 1 above, this means you can't assume APR format in your sorting.  In fact, you couldn't anyway because there are already existing artifacts in the production system that do not follow APR versioning.  In fact, versions don't even have to be numerical (e.g., `One-point-two` is technically valid).  So... you could make your sorting so it works correctly when comparing APR versions, but resorts to alphabetic otherwise.

3. I noticed that default sort order is now ascending by name.  I think we want to retain the original sort order, which was descending by last changed date.

4. I also noticed that the edit/delete buttons on the RHS are stacking vertically in most browser widths.  This is likely because of the addition of the new Date Created column.  Is it possible to get them to align side-by-side again?
image

Thanks again for taking this on!

**Comment by jsthomas1** [comment]
I was also thinking a few more things that are related: 1. Duplication is a feature we’ll want on the table (MedStar) 2. I think that we should  also have a dedicated button for viewing the artifact in the workspace rather that simply clicking the name of the artifact and 3. With even one more additional button, do you think that it would be preferable to wrap those three/four options into a context menu (either selectively based on screen width, or always)?

Perhaps something like [this](https://material-ui.com/components/menus/#CustomizedMenus.js) would be applicable, or another example of ctx menus on this page would. (Pending further discussion / refactor)

File: frontend/src/utils/artifactSort.js
Comment (kjmahalingam) [inline frontend/src/utils/artifactSort.js:21]: Stylistic point, but would you mind changing this to `aprRegex` instead? Snake case isn't very typical in our code, so camel case would be preferable.

File: frontend/src/utils/artifactSort.js
Comment (kjmahalingam) [inline frontend/src/utils/artifactSort.js:24]: Same comment as above, if you could change the casing for these that would be great.

File: frontend/src/components/artifact/ArtifactTable.js
Comment (kjmahalingam) [inline frontend/src/components/artifact/ArtifactTable.js:17]: Another mostly stylistic comment, but could you update this to simply assign the values of the array instead of pushing them all afterward?

**Review by kjmahalingam — COMMENTED** [review]
This looks awesome to me! I do have a few things that I noticed that I've made comments about inline.

In addition, I forgot to tell you how to run the tests, so sorry about that! You need to either run `yarn run lint` and `yarn test` in both the `frontend` and `api` directories, or you can go to the base directory and run the `./run-tests` script instead since it does all of those things.

Unfortunately since I forgot to tell you that, it looks like there are just a few tests breaking associated with these changes in both the `Artifact` and `ArtifactTable` tests. If you need help understanding how the tests work or why they're breaking, please feel free to reach out and we can help!

File: frontend/src/utils/artifactSort.js
Comment (jsthomas1) [reply frontend/src/utils/artifactSort.js:21]: You got it. I'm afraid of snakes myself.

**Comment by jsthomas1** [comment]
Broken tests are fixed. New ones for sort functionality in the works.

**Review by noranda — CHANGES_REQUESTED** [review]
Nice job on your first PR! React is fun and Material UI can be tricky, so I'm impressed you got this done so quickly. 🙌

I suggested some changes in the code, but most are stylistic or React best-practices that I suggested mostly so you can learn from them and to keep the code base consistent. Running `prettier` on these files would also fix a lot of the minor issues I found.

Functionally, the sorting works great! We will eventually want to add sorting on our other tables, but I suggest we wait until they are refactored before doing so. We will likely create a shared component to handle all the logic so we don't duplicate code.

My only other suggestion for this PR is to add some tests to make sure the sorting is behaving as expected. I didn't do an exhaustive check on the sorting, so some testing would help. If you need some help with the tests, just ask Dylan or myself.

Again, nice job!

File: frontend/src/components/artifact/ArtifactTableRow.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTableRow.js]: fixed

File: frontend/src/components/artifact/ArtifactTableRow.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTableRow.js]: fixed

File: frontend/src/components/artifact/ArtifactTableRow.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTableRow.js]: fixed

File: frontend/src/components/artifact/ArtifactTableRow.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTableRow.js]: fixed

File: frontend/src/components/artifact/ArtifactTable.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTable.js]: fixed

File: frontend/src/components/artifact/ArtifactTable.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTable.js]: fixed

File: frontend/src/components/artifact/ArtifactTable.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTable.js]: fixed

File: frontend/src/components/artifact/ArtifactTable.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTable.js]: fixed

File: frontend/src/utils/artifactSort.js
Comment (jsthomas1) [reply frontend/src/utils/artifactSort.js]: fixed

File: frontend/src/components/artifact/ArtifactTable.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTable.js]: fixed

File: frontend/src/utils/artifactSort.js
Comment (jsthomas1) [reply frontend/src/utils/artifactSort.js]: fixed

**Review by jsthomas1 — COMMENTED** [review]
These changes have been implemented alongside the changes of @noranda @kjmahalingam 
Should be good to merge (tests implemented and passed)

**Review by kjmahalingam — COMMENTED** [review]
Awesome! Looks great to me. Just a few *super* minor things.

1. I made some in-line comments about removing some extraneous logging statements.

2. I also noticed that if you `yarn run lint` in the `frontend` we actually do fail on line length for a couple of lines in the artifact sorting tests. I guess prettier doesn't catch everything our linter requires! Could you fix those as well?

3. For some reason, looks like the delete button is smaller than the edit button for me. I'll send you a picture of what it looks like.

Once those are done, consider this approved from me!

File: frontend/src/utils/artifactSort.js
Comment (jsthomas1) [reply frontend/src/utils/artifactSort.js]: resolved

File: frontend/src/utils/__tests__/artifactSort.test.js
Comment (jsthomas1) [reply frontend/src/utils/__tests__/artifactSort.test.js:122]: resolved

**Review by kjmahalingam — APPROVED** [review]
Good work! I think we have to decide on how to get this merged in once we have other approvals, but I'll leave it here for now!

File: frontend/src/components/artifact/ArtifactTable.js
Comment (noranda) [reply frontend/src/components/artifact/ArtifactTable.js:10]: Looks like this change didn't get in to your latest commits, can we change the names of these to be more explicit?

File: frontend/src/components/artifact/ArtifactTable.js
Comment (noranda) [reply frontend/src/components/artifact/ArtifactTable.js]: You changed the value to a boolean, which is great, but can you also change the names to be more explicit?

File: frontend/src/components/artifact/ArtifactTable.js
Comment (noranda) [reply frontend/src/components/artifact/ArtifactTable.js]: Can you take a look at these changes as well? I think they also got missed in your latest commits.

File: frontend/src/components/artifact/ArtifactTable.js
Comment (noranda) [inline frontend/src/components/artifact/ArtifactTable.js:20]: Can we change the name of this function to be `handleRequestSort` to be more explicit?

**Comment by noranda** [comment]
Looking better! Just a few issues left:

1. In `ArtifactTable.js`, line 10, can we change the names to make this more explicit?
```js
const [selectedColumnIndex, setSelectedColumnIndex] = useState(2);
```

2. In `ArtifactTable.js`, line 11, you changed this to a boolean, which is great, but can we also change the names to make this more explicit?
```js
const [sortAsc, setSortAsc] = useState(true);
```

3. In `ArtifactTable.js`, line 20, can you change the name of the function to `handleRequestSort` to be more explicit? You can also simplify this function a bit as follows.
```js
const handleRequestSort = columnIndex => {
  setSelectedColumnIndex(columnIndex);
  setSortAsc(columnIndex === selectedColumnIndex ? !sortAsc : true);
};
```

4.  In `ArtifactTable.js`, line 39, I had made the following comment, but I think Github may have collapsed it so you probably missed it:
 

> In React, if you're only returning JSX, you can actually remove the `return` and the curly brackets to tighten up the code a bit.
> 
> I would also get rid of the `padding="none"` attribute on the `TableCell` as I think we need the padding there. You can add `style={{ whiteSpace: 'nowrap' }}` to deal with the wrapping issue temporarily as discussed.
> 
> Finally, you can update the `TableSortLabel` attributes as follows if you updated to the `sortAsc` boolean and removed the `isActiveColumn` function.
> 
> ```suggestion
>             {columns.map((column, index) => (
>               <TableCell key={index} style={{ whiteSpace: 'nowrap' }}>
>                 <TableSortLabel
>                   id={index}
>                   direction={selectedColumnIndex !== index || sortAsc ? 'asc' : 'desc'}
>                   active={selectedColumnIndex === index}
>                   onClick={() => handleRequestSort(index)}
>                 >
>                   {column.columnName}
>                 </TableSortLabel>
>               </TableCell>
>             ))}
> ```

Thanks for adding tests for the sorting!

File: frontend/src/components/artifact/ArtifactTable.js
Comment (jsthomas1) [reply frontend/src/components/artifact/ArtifactTable.js]: resolved

**Review by noranda — APPROVED** [review]
This looks great! Nice work!

**Review by cmoesel — APPROVED** [review]
Thank you, Jacob!  Well done!

**Comment by cmoesel** [comment]
Closing this as it was merged as squashed commit 24b717b.  Thanks again, @jsthomas1!

**Pull request closed without merging**
